### PR TITLE
[Large 2D load for GEMM - #7] Enable the large size 2d load in AccelerateMatmul pass.

### PR DESCRIPTION
**Pull request by chengjunlu**
Use DPAS layout with the size > 1 of repCluster in AccelerateMatmul pass when possible.

### CODE REVIEW

File: third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp [hunk 1]
```diff
@@ -150,6 +150,33 @@ class BlockedToDPAS : public RewritePattern {
...
+        repClusterDimM = 1;
+      }
+
+      unsigned repClusterDimN;
+      auto repB = dpasEnc.getDPASRepetitions(oldBType.getShape(), 1);
+      if (repB[1] >= 2) {
+        repClusterDimN = 2;
+      } else {
+        repClusterDimN = 1;
+      }
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp:172]:
```suggestion
unsigned repClusterDimN = (repB[1] >= 2) ? 2 : 1;
```

Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp:172]:
```suggestion
      SmallVector<int64_t> repA =
          dpasEnc.getDPASRepetitions(oldAType.getShape(), 0);
      unsigned repClusterDimM = std::min(4u, static_cast<unsigned>(repA[0]));

      SmallVector<int64_t> repB =
          dpasEnc.getDPASRepetitions(oldBType.getShape(), 1);
      unsigned repClusterDimN = std::min(2u, static_cast<unsigned>(repB[1]));
```

File: test/TritonIntelGPU/accelerate-matmul-pvc.mlir [hunk 2]
```diff
@@ -131,3 +132,60 @@ module attributes {"triton_gpu.target" = "xpu:DEVICE_ARCH.PVC", "triton_gpu.num-
...
+    %result = tt.dot %a, %b, %zero_f32 : tensor<128x128xf16, #triton_gpu.dot_op<{opIdx = 0, parent = #blocked}>> * tensor<128x128xf16, #triton_gpu.dot_op<{opIdx = 1, parent = #blocked}>> -> tensor<128x128xf32, #blocked>
+    %result_ptr = tt.splat %arg0 : !tt.ptr<f32> -> tensor<128x128x!tt.ptr<f32>, #blocked>
+    tt.store %result_ptr, %result, %mask : tensor<128x128x!tt.ptr<f32>, #blocked>
+    tt.return
+  }
+}
+
+// -----
+
+// CHECK: #mma = #triton_intel_gpu.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [1, 1], repCluster = [2, 2], A = [16, 16], B = [16, 32], C = [16, 32]}>
```
Comment (etiotto) [inline test/TritonIntelGPU/accelerate-matmul-pvc.mlir:157]: These next 2 tests do not have any CHECK in the function body. Can we add some key CHECKS ?

**Review by etiotto — APPROVED** [review]

File: third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp [hunk 3]
```diff
@@ -152,6 +152,23 @@ class BlockedToDPAS : public RewritePattern {
...
+    if (arch == intel::DeviceArch::PVC) {
+      // Enlarge the repCluster size to use the large 2D load for A and B
+      // operands.
+      SmallVector<int64_t> repA =
+          dpasEnc.getDPASRepetitions(oldAType.getShape(), 0);
+      unsigned repClusterDimM = std::min(4u, static_cast<unsigned>(repA[0]));
+
+      SmallVector<int64_t> repB =
+          dpasEnc.getDPASRepetitions(oldBType.getShape(), 1);
+      unsigned repClusterDimN = std::min(2u, static_cast<unsigned>(repB[1]));
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp:164]: Can we name `4u` and `2u` here to see where they come from?

Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp:164]: @chengjunlu FYA ^^^

Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp:164]: The maximum capability of the 2D load is: 64 bytes per row and 32 rows, total 2k bytes.

The properties of the operands A to the DPAS instruction: 32 bytes per row and 8 rows (for all type). So if the number of repetitions per warp is larger than the `4u` on the M dimension, we want to cluster the 4 repetitions then use one 2D load instruction to load 4x8 rows operands A.

The properties of the operands B to the DPAS instruction: 32 bytes per row and 16 rows (for fp16/bf16). So if the number of repetitions per warp is larger than the `2u` on the N dimension, we want to cluster the 2 repetitions then use one 2D load instruction to load 2x32 bytes operands B. 

I will add the changes:
- Name the constant value with proper name.
- I missed the case for the int8 type for DPAS. Will add the int8 support too.

**Review by whitneywhtsang — APPROVED** [review]

File: third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp [hunk 4]
```diff
@@ -10,6 +10,9 @@
 #include "triton/Dialect/TritonGPU/IR/Dialect.h"
 #include "llvm/ADT/TypeSwitch.h"
 
+#define PVC_2D_LOAD_MAXIMUM_NUMBER_OF_ROWS 32
+#define PVC_2D_LOAD_MAXIMUM_BYTES_OF_COLS 64
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp:14]:
```suggestion
static constexpr unsigned pvc2DLoadMaximumNumberOfRows = 32;
static constexpr unsigned pvc2DLoadMaximumBytesOfCols = 64;
```

What does "bytes of cols" mean?

File: third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp [hunk 5]
```diff
@@ -149,6 +152,30 @@ class BlockedToDPAS : public RewritePattern {
         dpasCap.executionSize, opsPerChan, warpsPerTile, {1, 1},
         threadsPerWarp);
 
+    if (dpasCap.executionSize == 16 /* PVC */) {
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp:155]: Can we make 16 a constant somewhere? Also, is this branch only reachable on PVC arch?

File: third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp (hunk 4 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp:14]: Total number of bytes per row. Does `TotalNumberBytesPerRow` make more sense?

File: third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp (hunk 5 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp:155]: This branch should only reachable on PVC device so far.

File: third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp (hunk 4 above)
Comment (victor-eds) [reply third_party/intel/lib/TritonIntelGPUTransforms/AccelerateMatmul.cpp:14]: Yeah, I think so!

**Review by victor-eds — APPROVED** [review]

**Pull request merged**